Test combined infra metrics

arnout opened the pull request:
A number of fixes are required along the way.

arnout committed:
agent: slave: handle_ap_metrics_query: correctly parse TLVs

This function escaped from the addClass -> getClass conversion because
it incorrectly used cmdu_tx instead of cmdu_rx. And it is not used in
test_flows. Therefore, its implementation was removed completely.

Re-introduce the parsing, but now properly with getClass. We expect
exactly one ApMetricQuery TLV for this message.

While we're at it, also log the bssids.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
controller: fix type AP_METRIC_RESPONSE -> METRICS

This is annoying when grepping for that string.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tests: test_flows: properly check exit value of send_CAPI_command

The result of send_CAPI_command is not checked, so failures to send it
are not registered as failures.

In shell it is quite difficult to get the exit code of a $() subshell,
so work around this by redirecting to a temporary file instead. Getting
the exit code of the first step in a pipeline is equally difficult, so
move the pipe to where the result is used.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tests: test_flows: add send_CAPI_1905 function

One of the most often used CAPI commands is dev_send_1905, so let's make
a function to simplify that one.

While we're at it, store the MID that we get as return value in the
capi_command_mid variable. Use this in the one place where the MID is
checked.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
agent: backhaul_manager: handle combined infra metrics

The backhaul manager is already subscribed to the combined infra metrics
message. Handle the message simply by logging it, and sending an ACK.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
controller: master_thread: always construct combined_infra_metrics

The construct_combined_infra_metric function checks if all the agents in
the database have sent their link metrics before the message is sent.
However, there is really no reason to wait for that. Maybe not all
agents will send link metrics, and anyway we're just filling a
certification buffer which will only be flushed when explicitly asked by
the UCC.

Always construct the combined infra metrics, even if not all link
metrics have been received.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
controller: master_thread: construct combined_infra_metrics also for AP Metrics

The master thread constructs the combined_infra_metric message when a
link metrics has been received. That is because the certification test
first requests AP Metrics and then Link Metrics. However, in reality
this can be done in any order, so let's construct it also when AP
Metrics are received.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

tomereli commented on tests/test_flows.sh:
```diff
@@ -84,6 +84,19 @@ send_CAPI_command() {
+    for arg; do
+        command="$command,$arg"
+    done
+    capi_command_mid=""
```
Why is this needed? you set this 2 lines below

arnout replied:
(For future reference) It is needed because of the `return $?` in the following line, to make sure the mid is not set if the command fails.

tomereli commented on tests/test_flows.sh:
```diff
@@ -207,35 +220,63 @@ test_client_capability_query() {
 test_combined_infra_metrics() {
-    err "combined_infra_metrics not implemented yet."
-    return 0
+    status "test combined infra metrics"
```
This has to be in a separate commit

tomereli reviewed (changes requested):
Looking good so far, but the test MUST make sure the agents actually sends the link metrics and check that the controller adds these as TLVs to the combined infrastructure message. This is the bug that is fixed by #669 so we have to be able to catch these kind of bugs in the test.

YuriyMasechko committed:
test_flows.sh: add dummy values to combined infrastructure metrics

Add dummy values to link metrics and ap metrics queries for
combined infrastructure metrics test.

Signed-off-by: Yuriy Masechko <yuriy.masechko@globallogic.com>

tomereli reviewed (changes requested):
> Looking good so far, but the test MUST make sure the agents actually sends the link metrics and check that the controller adds these as TLVs to the combined infrastructure message. This is the bug that is fixed by #669 so we have to be able to catch these kind of bugs in the test.

@YuriyMasechko I don't see that this is taken care of by this PR. The problem is that when the UCC requests to send an empty combined infrastructure metrics like you did here:
```
# Trigger combined infra metrics
dbg "Send Combined infrastructure metrics message to agent 1"
check send_CAPI_1905 ${GATEWAY} $mac_agent1 0x8013
```

The UCC listener should send the preset certification CMDU. If there is no such CMDU (which was the bug solved in my other PR), then the UCC listener will send an *empty* 1905 CMDU with message type 0x8013 - the test failed in the certification since the CMDU was empty, while it should have contained TLVs.

In this case, I expect the test to simulate the same thing and actually verify that the received CMDU in the agent contains the TLVs we expect.
This is why I suggested to add dummy link metrics for now in the agent, so we can verify that the combined infrastructure metrics include these same values and is not just an empty CMDU which is a bug.

YuriyMasechko commented:
> > Looking good so far, but the test MUST make sure the agents actually sends the link metrics and check that the controller adds these as TLVs to the combined infrastructure message. This is the bug that is fixed by #669 so we have to be able to catch these kind of bugs in the test.
> 
> @YuriyMasechko I don't see that this is taken care of by this PR. The problem is that when the UCC requests to send an empty combined infrastructure metrics like you did here:
> 
> ```
> # Trigger combined infra metrics
> dbg "Send Combined infrastructure metrics message to agent 1"
> check send_CAPI_1905 ${GATEWAY} $mac_agent1 0x8013
> ```
> 
> The UCC listener should send the preset certification CMDU. If there is no such CMDU (which was the bug solved in my other PR), then the UCC listener will send an _empty_ 1905 CMDU with message type 0x8013 - the test failed in the certification since the CMDU was empty, while it should have contained TLVs.
> 
> In this case, I expect the test to simulate the same thing and actually verify that the received CMDU in the agent contains the TLVs we expect.
> This is why I suggested to add dummy link metrics for now in the agent, so we can verify that the combined infrastructure metrics include these same values and is not just an empty CMDU which is a bug.

Added check for Transmitter/Receiver link metrics TLV inside handle_1905_combined_infrastructure_metrics() and test_flows.
